Switch to Whoosh implementation to improve speed

eepMoody opened the pull request:
Replaces the existing hand-tuned search with the Whoosh framework.

This has major performance benefits, but incurs a much higher indexing time. To deal with this, I've adopted a strategy that caches indexes in the repo, unpacks them for quick development purposes, then runs a background indexing process inside the server after deployment. In practice, this means search results may start slightly out-of-date (if a major content release has not been cachced) but will be consistent within approximately 5 minutes of deploy.

Major changes:

- Replace hand-tuned and ES options with Whoosh
- replace vector search with spaCy embeddings for true semantic matching
- add management commands, including `quickindex` which will index and create zipped files for checking in
- add tests covering all search methods, ensuring they return realistic results
- add background process in server to index live content in production


NOTE being kept in draft until #844 is merged, as this is based on that branch

augustjohnson commented:
@eepMoody what's the status on this PR? If I understand correctly, a fair amount of the search is already in place in staging... Where does this fit in?

eepMoody commented:
Ah sorry, I hit a bit of a wall on getting the actual switch to Elasticsearch (which is a major performance boost) sorted out. What's in production adds the new search methods, but the performance isn't really improved. Might be a little bit worse, but hard to tell until it hits real resources.

The main bit that's outstanding is the deployment setup, since ES needs to run in its own container/server. I haven't quite gotten that into a state where it's functional in production.

If you want to take a peek, maybe there's something obvious that I've overlooked or another approach that might make more sense?

eepMoody committed:
Fix spell data regressions from PR #842 and update newrelic API

eepMoody committed:
add missing higher-leve options

eepMoody committed:
add tests for duplicate or missing higher_level values

eepMoody committed:
Add missing casting options for deepm and spells-that-dont-suck

eepMoody committed:
fix freezing sphere being merged with freedom of movement

eepMoody committed:
load data properly, dial in weighting

eepMoody committed:
improve highlighting

eepMoody committed:
update build to load ml dependencies after start

eepMoody committed:
setup docker config for elasticsearch server

eepMoody committed:
add core elasticsearch solution

eepMoody committed:
rework to use scikit instead of torch for smaller build

eepMoody committed:
more docker fixes

eepMoody committed:
docker build fixes

eepMoody committed:
update start script to properly start ES server in DO

eepMoody committed:
modify dockerfile

eepMoody committed:
Update Dockerfile to create necessary Elasticsearch directories and set permissions

eepMoody committed:
enterprise stuff again

eepMoody committed:
attempt different build strategy for DO

eepMoody committed:
missing comma

eepMoody committed:
update instead of append Elasticsearch configuration settings

eepMoody committed:
Update app.yaml to improve deployement process

eepMoody committed:
switch ES to use /tmp directories since we are ephemeral

eepMoody committed:
Add auto-generation of SECRET_KEY in start.sh

eepMoody committed:
cleanup DO config

eepMoody committed:
go back to docker

eepMoody committed:
Remove enterprise stuff that doesn't work in DO

eepMoody committed:
Refactor Dockerfile to use GPG keyring for Elasticsearch installation

eepMoody committed:
Pre-build and commit search indexes

eepMoody committed:
update docs

eepMoody committed:
cleanup scripts

eepMoody committed:
Replace Elasticsearch with Whoosh and spaCy for in-process search

Use Whoosh for text/fuzzy search and spaCy word embeddings for
semantic similarity. Removes the need for a separate ES service.

eepMoody committed:
Add search tests and fix fuzzy search

eepMoody committed:
clean up, lint

eepMoody committed:
add dummy secret key for build process
